py/objexcept: Make sure mp_obj_new_exception_msg doesn't copy/format msg

dpgeorge opened the pull request:
mp_obj_new_exception_msg() assumes that the message passed to it is in ROM
and so can use its data directly to create the string object for the
argument of the exception, saving RAM.  At the same time, this approach
also makes sure that there is no attempt to format the message with printf,
which could lead to faults if the message contained % characters.

Fixes issue #3004.

dpgeorge commented:
This has a small increase in code size:
```
   bare-arm:   +64 +0.097% 
minimal x86:  +184 +0.119% 
   unix x64:  +192 +0.039% 
unix nanbox:  +200 +0.046% 
      stm32:   +88 +0.024% 
     cc3200:   +80 +0.043% 
    esp8266:   +92 +0.014% 
      esp32:  +132 +0.012%
```
But that's probably worth it because now exception messages that don't have any formatting (eg something like `TypeError: exceptions must derive from BaseException`), of which there are 20 in py/ and 29 in the rest of the code base, don't allocate memory to copy the static message string to (which was the case with the old behaviour, which always assumed that the string needed to be formatted so called printf on it and copied it to a new, dynamically allocated buffer on the heap).

As a bonus, this RAM optimisation fixes #3004.

dpgeorge commented:
Note: Travis failure is only due to increase in code size of minimal port.

pfalcon commented on py/objexcept.c:
```diff
@@ -313,7 +313,36 @@ mp_obj_t mp_obj_new_exception_args(const mp_obj_type_t *exc_type, size_t n_args,
+    // that buffer to store the string object, reserving room at the start for the
+    // traceback and 1-tuple.
+    if (o_str == NULL
+        && mp_emergency_exception_buf_size >= EMG_TRACEBACK_ALLOC * sizeof(size_t)
```
This is of course hard to read and repeats the same (longish) subexpression twice. (And it's 2nd change in row like that, after adding similarly hardish to read code to implement accept queue in modlwip, where no attempt was made to brush it up somehow (hide behind macros/inline funs, whatever)).

stinos replied:
Indeed reusing this shouldn't be too hard and would do wonders for readability

pfalcon commented:
Looks good. It's actually comes as a surprise that there was no dedicated func which creates `str` from const str data. (2nd possible "surprise" is that `str` itself is still allocated on heap, but given that `Exception` itself is allocated anyway, apparently shouldn't be a surprise).

stinos commented:
> Fixes issue #3004.

Confirmed.

dpgeorge commented:
Thanks for the feedback and testing.

See #4338 for a PR to clean up the expressions used to calculate offsets into the emergency exception buffer.  This PR would eventually be rebased on that one and use its macros.

dpgeorge commented:
> of which there are 20 in py/ and 29 in the rest of the code base

Actually, `mp_raise_TypeError` and `mp_raise_ValueError` also use `mp_obj_new_exception_msg`, so that gives about 500 uses in the whole code base for this latter function, so optimising it to not need the heap seems worth the cost in code space.

dpgeorge committed:
py/objexcept: Make sure mp_obj_new_exception_msg doesn't copy/format msg

mp_obj_new_exception_msg() assumes that the message passed to it is in ROM
and so can use its data directly to create the string object for the
argument of the exception, saving RAM.  At the same time, this approach
also makes sure that there is no attempt to format the message with printf,
which could lead to faults if the message contained % characters.

Fixes issue #3004.

dpgeorge commented:
Merged in 55830dd9bf4fee87c0a6d3f38c51614fea0eb483
